File: divide_and_conquer/MergeSort/MergeSort.go
package main

import (
    "fmt"
)

// merge merges two sorted slices into a single sorted slice
func merge(left, right []int) []int {
    // Create a result slice with a capacity equal to the sum of the lengths of left and right slices
    result := make([]int, 0, len(left)+len(right))
    i, j := 0, 0

    // Compare elements from left and right slices and append the smaller one to the result
    for i < len(left) && j < len(right) {
        if left[i] < right[j]) {
            result = append(result, left[i])
            i++
        } else {
            result = append(result, right[j])
            j++
        }
    }

    // Append any remaining elements from the left slice
    result = append(result, left[i:]...)
    // Append any remaining elements from the right slice
    result = append(result, right[j:]...)

    return result
}

// mergeSort sorts a slice of integers using the merge sort algorithm
func mergeSort(arr []int) []int {
    if len(arr) <= 1 {
        return arr
    }

    // Find the middle point to divide the array into two halves
    mid := len(arr) / 2

    // Call mergeSort recursively on the first half
    left := mergeSort(arr[:mid])
    // Call mergeSort recursively on the second half
    right := mergeSort(arr[mid:])

    // Merge the two halves
    return merge(left, right)
}

func main() {
    arr := []int{12, 11, 13, 5, 6, 7}
    fmt.Println("Given array is:", arr)
    sortedArr := mergeSort(arr)
    fmt.Println("Sorted array is:", sortedArr)
}

File: divide_and_conquer/MergeSort/MergeSort.cpp
#include <iostream>
#include <vector>

// Function to merge two subarrays of arr[].
// First subarray is arr[left..mid]
// Second subarray is arr[mid+1..right]
void merge(std::vector<int>& arr, int left, int mid, int right) {
    int n1 = mid - left + 1; // Size of the first subarray
    int n2 = right - mid;    // Size of the second subarray

    // Create temporary arrays to hold the elements of the subarrays
    std::vector<int> L(n1);
    std::vector<int> R(n2);

    // Copy data to temporary arrays L[] and R[]
    for (int i = 0; i < n1; i++)
        L[i] = arr[left + i];
    for (int j = 0; j < n2; j++)
        R[j] = arr[mid + 1 + j];

    // Initial indexes of the first and second subarrays
    int i = 0, j = 0;
    // Initial index of the merged subarray
    int k = left;

    // Merge the temporary arrays back into arr[left..right]
    while (i < n1 && j < n2) {
        if (L[i] <= R[j]) {
            arr[k] = L[i];
            i++;
        } else {
            arr[k] = R[j];
            j++;
        }
        k++;
    }

    // Copy the remaining elements of L[], if there are any
    while (i < n1) {
        arr[k] = L[i];
        i++;
        k++;
    }

    // Copy the remaining elements of R[], if there are any
    while (j < n2) {
        arr[k] = R[j];
        j++;
        k++;
    }
}

// Function to implement merge sort
void mergeSort(std::vector<int>& arr, int left, int right) {
    if (left < right) {
        int mid = left + (right - left) / 2;

        // Sort first and second halves
        mergeSort(arr, left, mid);
        mergeSort(arr, mid + 1, right);

        // Merge the sorted halves
        merge(arr, left, mid, right);
    }
}

int main() {
    std::vector<int> arr = {12, 11, 13, 5, 6, 7};
    std::cout << "Given array is \n";
    for (int i = 0; i < arr.size(); i++)
        std::cout << arr[i] << " ";
    std::cout << std::endl;

    mergeSort(arr, 0, arr.size() - 1);

    std::cout << "\nSorted array is \n";
    for (int i = 0; i < arr.size(); i++)
        std::cout << arr[i] << " ";
    std::cout << std::endl;
    return 0;
}